lock-api/src/main/java/com/palantir/lock/client/MultiClientTransactionStarter.java

```diff
@@ -0,0 +1,323 @@
+            Map<Namespace, RequestParams> originalRequestMap,
+            InternalMultiClientConjureTimelockService delegate,
+            UUID requestorId) {
+
+        Map<Namespace, ConjureStartTransactionsRequest> namespaceWiseRequests =
+                getNamespaceWiseRequests(originalRequestMap, requestorId);
+        Map<Namespace, ConjureStartTransactionsResponse> responseMap = getResponseMap(delegate, namespaceWiseRequests);
+        return KeyedStream.stream(responseMap)
+                .mapEntries((namespace, response) -> {
+                    TransactionStarterHelper.updateCacheWithStartTransactionResponse(
+                            originalRequestMap.get(namespace).cache(),
+                            fromConjure(namespaceWiseRequests.get(namespace).getLastKnownVersion()),
+                            response);
+                    return Maps.immutableEntry(namespace, TransactionStarterHelper.split(response));
+                })
+                .collectToMap();
+    }
+
+    @VisibleForTesting
+    static Map<Namespace, ConjureStartTransactionsRequest> getNamespaceWiseRequests(
+            Map<Namespace, RequestParams> originalRequestMap, UUID requestorId) {
+        return KeyedStream.stream(originalRequestMap)
+                .mapEntries((namespace, requestParams) ->
+                        Maps.immutableEntry(namespace, getConjureRequest(requestParams, requestorId)))
+                .collectToMap();
+    }
+
+    private static ConjureStartTransactionsRequest getConjureRequest(RequestParams requestParams, UUID requestorId) {
+        return ConjureStartTransactionsRequest.builder()
+                .requestorId(requestorId)
+                .requestId(UUID.randomUUID())
+                .numTransactions(requestParams.numTransactions())
+                .lastKnownVersion(toConjure(requestParams.cache().lastKnownVersion()))
+                .build();
+    }
+
+    private static Map<Namespace, ConjureStartTransactionsResponse> getResponseMap(
+            InternalMultiClientConjureTimelockService delegate,
+            Map<Namespace, ConjureStartTransactionsRequest> namespaceWiseRequests) {
+        return KeyedStream.stream(delegate.startTransactions(namespaceWiseRequests))
+                .map(LockLeaseService::assignLeasedLockTokenToImmutableTimestampLock)
+                .collectToMap();
+    }
+
+    @Override
+    public void close() {
+        autobatcher.close();
+    }
+
+    @Value.Immutable
+    interface NamespaceAndRequestParams {
+        @Value.Parameter
+        Namespace namespace();
+
+        @Value.Parameter
+        RequestParams params();
+
+        static NamespaceAndRequestParams of(Namespace namespace, RequestParams params) {
+            return ImmutableNamespaceAndRequestParams.of(namespace, params);
+        }
+    }
+
+    @Value.Immutable
+    interface RequestParams {
+        @Value.Parameter
+        Integer numTransactions();
```

Comment:
This class is basically always used in a way where this number changes and the other two params are fixed. Given we don't really use the immutability or the generated code anywhere, maybe this should just be a normal concrete class with a method to change `numTransactions`? That said, we are not creating millions of objects, so it'a also fine to leave as is

Reply:
Right, but we will need need to do one iteration of cloning as the params in the original BatchElement should not be updated, which will become one more thing to worry about in this class.

Reply:
Note - We actually do use the immutability of this class to ensure that the request batch cannot have its state mutated.